File: core/src/main/java/at/stefangeyer/challonge/service/implementation/SimpleAttachmentService.java
package at.stefangeyer.challonge.service.implementation;

import at.stefangeyer.challonge.async.Callback;
import at.stefangeyer.challonge.exception.DataAccessException;
import at.stefangeyer.challonge.model.Attachment;
import at.stefangeyer.challonge.model.Match;
import at.stefangeyer.challonge.model.query.AttachmentQuery;
import at.stefangeyer.challonge.model.wrapper.AttachmentWrapper;
import at.stefangeyer.challonge.rest.AttachmentRestClient;
import at.stefangeyer.challonge.service.AttachmentService;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Attachment Service Implementation
 *
 * @author Stefan Geyer
 * @version 2018-06-30
 */
public class SimpleAttachmentService implements AttachmentService {

    private AttachmentRestClient restClient;

    public SimpleAttachmentService(AttachmentRestClient restClient) {
        this.restClient = restClient;
    }

    @Override
    public List<Attachment> getAttachments(Match match) throws DataAccessException {
        return this.restClient.getAttachments(String.valueOf(match.getTournamentId()), match.getId()).stream()
                .map(AttachmentWrapper::getMatchAttachment).collect(Collectors.toList());
    }

    @Override
    public void getAttachments(Match match, Callback<List<Attachment>> onSuccess,
                               Callback<DataAccessException> onFailure) {
        this.restClient.getAttachments(String.valueOf(match.getTournamentId()), match.getId(),
                list -> onSuccess.accept(list.stream().map(AttachmentWrapper::getMatchAttachment)
                        .collect(Collectors.toList())), onFailure);
    }

    @Override
    public Attachment getAttachment(Match match, long attachmentId) throws DataAccessException {
        return this.restClient.getAttachment(String.valueOf(match.getTournamentId()), match.getId(),
                attachmentId).getMatchAttachment();
    }

    @Override
    public void getAttachment(Match match, long attachmentId, Callback<Attachment> onSuccess,
                              Callback<DataAccessException> onFailure) {
        this.restClient.getAttachment(String.valueOf(match.getTournamentId()), match.getId(), attachmentId,
                aw -> onSuccess.accept(aw.getMatchAttachment()), onFailure);
    }

    @Override
    public Attachment createAttachment(Match match, AttachmentQuery data) throws DataAccessException {
        validateAttachmentQuery(data);

        return this.restClient.createAttachment(String.valueOf(match.getTournamentId()), match.getId(),
                data).getMatchAttachment();
    }

    @Override
    public void createAttachment(Match match, AttachmentQuery data, Callback<Attachment> onSuccess,
                                 Callback<DataAccessException> onFailure) {
        validateAttachmentQuery(data);

        this.restClient.createAttachment(String.valueOf(match.getTournamentId()), match.getId(), data,
                aw -> onSuccess.accept(aw.getMatchAttachment()), onFailure);
    }

    @Override
    public Attachment updateAttachment(Match match, Attachment attachment,
                                       AttachmentQuery data) throws DataAccessException {
        validateAttachmentQuery(data);

        return this.restClient.updateAttachment(String.valueOf(match.getTournamentId()), match.getId(),
                attachment.getId(), data).getMatchAttachment();
    }

    @Override
    public void updateAttachment(Match match, Attachment attachment, AttachmentQuery data,
                                 Callback<Attachment> onSuccess, Callback<DataAccessException> onFailure) {
        validateAttachmentQuery(data);

        this.restClient.updateAttachment(String.valueOf(match.getTournamentId()), match.getId(),
                attachment.getId(), data, aw -> onSuccess.accept(aw.getMatchAttachment()), onFailure);
    }

    @Override
    public Attachment deleteAttachment(Match match, Attachment attachment) throws DataAccessException {
        return this.restClient.deleteAttachment(String.valueOf(match.getTournamentId()), match.getId(),
                attachment.getId()).getMatchAttachment();
    }

    @Override
    public void deleteAttachment(Match match, Attachment attachment, Callback<Attachment> onSuccess,
                                 Callback<DataAccessException> onFailure) {
        this.restClient.deleteAttachment(String.valueOf(match.getTournamentId()), match.getId(),
                attachment.getId(), aw -> onSuccess.accept(aw.getMatchAttachment()), onFailure);
    }

    private void validateAttachmentQuery(AttachmentQuery data) {
        if (data.getAsset() == null && data.getUrl() == null && data.getDescription() == null) {
            throw new IllegalArgumentException("All data parameters are null. Provide at least one");
        }
    }
}

File: src/main/kotlin/at/stefangeyer/challonge/service/implementation/SimpleAttachmentService.kt
package at.stefangeyer.challonge.service.implementation

import at.stefangeyer.challonge.async.Callback
import at.stefangeyer.challonge.exception.DataAccessException
import at.stefangeyer.challonge.model.Attachment
import at.stefangeyer.challonge.model.Match
import at.stefangeyer.challonge.model.query.AttachmentQuery
import at.stefangeyer.challonge.rest.AttachmentRestClient
import at.stefangeyer.challonge.service.AttachmentService

/**
 * Attachment Service Implementation
 *
 * @author Stefan Geyer
 * @version 2018-06-30
 */
class SimpleAttachmentService(private val restClient: AttachmentRestClient) : AttachmentService {

    override fun getAttachments(match: Match): List<Attachment> =
            this.restClient.getAttachments(match.tournamentId.toString(), match.id).map { aw -> aw.attachment }

    override fun getAttachments(match: Match,
                                onSuccess: Callback<List<Attachment>>, onFailure: Callback<DataAccessException>) {
        this.restClient.getAttachments(match.tournamentId.toString(), match.id,
                { list -> onSuccess(list.map { aw -> aw.attachment }) }, onFailure)
    }

    override fun getAttachment(match: Match, attachmentId: Long): Attachment =
            this.restClient.getAttachment(match.tournamentId.toString(), match.id, attachmentId).attachment

    override fun getAttachment(match: Match, attachmentId: Long,
                               onSuccess: Callback<Attachment>, onFailure: Callback<DataAccessException>) {
        this.restClient.getAttachment(match.tournamentId.toString(), match.id, attachmentId,
                { aw -> onSuccess(aw.attachment) }, onFailure)
    }

    override fun createAttachment(match: Match, data: AttachmentQuery): Attachment {
        validateAttachmentQuery(data)
        return this.restClient.createAttachment(match.tournamentId.toString(), match.id, data).attachment
    }

    override fun createAttachment(match: Match, data: AttachmentQuery,
                                  onSuccess: Callback<Attachment>, onFailure: Callback<DataAccessException>) {
        validateAttachmentQuery(data)
        return this.restClient.createAttachment(match.tournamentId.toString(), match.id, data,
                { aw -> onSuccess(aw.attachment) }, onFailure)
    }

    override fun updateAttachment(match: Match, attachment: Attachment, data: AttachmentQuery): Attachment {
        validateAttachmentQuery(data)
        return this.restClient.updateAttachment(match.tournamentId.toString(), match.id, attachment.id, data).attachment
    }

    override fun updateAttachment(match: Match, attachment: Attachment, data: AttachmentQuery,
                                  onSuccess: Callback<Attachment>, onFailure: Callback<DataAccessException>) {
        validateAttachmentQuery(data)
        return this.restClient.updateAttachment(match.tournamentId.toString(), match.id, attachment.id, data,
                { aw -> onSuccess(aw.attachment) }, onFailure)
    }

    override fun deleteAttachment(match: Match, attachment: Attachment): Attachment =
            this.restClient.deleteAttachment(match.tournamentId.toString(), match.id, attachment.id).attachment

    override fun deleteAttachment(match: Match, attachment: Attachment,
                                  onSuccess: Callback<Attachment>, onFailure: Callback<DataAccessException>) {
        this.restClient.deleteAttachment(match.tournamentId.toString(), match.id, attachment.id,
                { aw -> onSuccess(aw.attachment) }, onFailure)
    }

    private fun validateAttachmentQuery(data: AttachmentQuery) {
        if (data.asset == null && data.url == null && data.description == null) {
            throw IllegalArgumentException("All data parameters are null. Provide at least one")
        }
    }
}